File: csharp_component/component/dbproxy/hubmanager.cs
using System;
using System.Collections;
using System.Collections.Generic;

namespace DBProxy
{
	public class HubManager
	{
		private Dictionary<string, HubProxy> hubproxys_name;
		private Dictionary<Abelkhan.Ichannel, HubProxy> hubproxys;

		private List<string> closed_hub_list = new ();

		public HubManager()
		{
			hubproxys_name = new Dictionary<string, HubProxy> ();
			hubproxys = new Dictionary<Abelkhan.Ichannel, HubProxy> ();
		}

		public HubProxy reg_hub(Abelkhan.Ichannel ch, String name)
		{
			HubProxy _hubproxy = new HubProxy (ch);
			if (hubproxys_name.TryGetValue(name, out HubProxy _old_proxy))
            {
				hubproxys_name[name] = _hubproxy;
			}
            else
            {
				hubproxys_name.Add(name, _hubproxy);
			}
			hubproxys[ch] = _hubproxy;

			return _hubproxy;
		}

		public void on_hub_closed(string name)
		{
			if (!closed_hub_list.Contains(name))
			{
				closed_hub_list.Add(name);
			}
        }

		public bool get_hub(Abelkhan.Ichannel ch, out HubProxy _proxy)
		{
            return hubproxys.TryGetValue(ch, out _proxy);
		}

		public bool all_hub_closed()
        {
			return hubproxys_name.Count == closed_hub_list.Count;
		}

	}
}

File: cpp_component/component/dbproxy/hubmanager.cpp
/*
 * hubmanager.cpp
 * qianqians
 * 2024/5/30
 */
#include "hubmanager.h"

namespace dbproxy
{

hubmanager::hubmanager() {
}

std::shared_ptr<hubproxy> hubmanager::reg_hub(std::shared_ptr<abelkhan::Ichannel> ch, std::string name)
{
	auto _hubproxy = std::make_shared<hubproxy>(ch);
	hubproxys_name[name] = _hubproxy;
	hubproxys[ch] = _hubproxy;

	return _hubproxy;
}

void hubmanager::on_hub_closed(std::string name)
{
	if (std::find(closed_hub_list.begin(), closed_hub_list.end(), name) == closed_hub_list.end())
	{
		closed_hub_list.push_back(name);
	}
}

bool hubmanager::get_hub(std::shared_ptr<abelkhan::Ichannel> ch, std::shared_ptr<hubproxy>& _proxy)
{
	auto it = hubproxys.find(ch);
	if (it != hubproxys.end()) {
		_proxy = it->second;
		return true;
	}

	_proxy = nullptr;
	return false;
}

bool hubmanager::all_hub_closed()
{
	return hubproxys_name.size() <= closed_hub_list.size();
}

}